snapcraft/internal/sources/_git.py

```diff
@@ -49,14 +53,18 @@ def _pull_existing(self):
         elif self.source_commit:
             refspec = self.source_commit
 
-        # Pull changes to this repository and any submodules.
+        reset_spec = refspec if refspec != 'HEAD' else 'origin/master'
+
+        subprocess.check_call([self.command, '-C', self.source_dir,
+                               'fetch', '--prune',
+                               '--recurse-submodules=yes'], **self.kwargs)
         subprocess.check_call([self.command, '-C', self.source_dir,
-                               'pull', '--recurse-submodules=yes',
-                               self.source, refspec])
+                               'reset', '--hard', reset_spec], **self.kwargs)
 
         # Merge any updates for the submodules (if any).
         subprocess.check_call([self.command, '-C', self.source_dir,
-                              'submodule', 'update'])
+                              'submodule', 'update', '--remote'],
+                              **self.kwargs)
```

Comment:
Should this include `--recursive` to recursively make sure submodules within submodules are up-to-date?